python/shotgun_fields/bubble_widget.py

```diff
@@ -0,0 +1,495 @@
+        format.setObjectType(self._bubble_text_object.OBJECT_TYPE)
+        format.setProperty(self._bubble_text_object.BUBBLE_DATA_PROPERTY, bubble_id)
+
+        # keep a reference to the format so that we can map a cursor to this
+        # bubble later on
+        self._formats[bubble_id] = format
+
+        bubble.remove_clicked.connect(
+            lambda: self.remove_bubble(bubble_id)
+        )
+
+        # insert the bubble character into the text editor and format it properly
+        cursor = self.textCursor()
+        cursor.beginEditBlock()
+        cursor.insertText(self._OBJECT_REPLACEMENT_CHAR, format)
+        cursor.endEditBlock()
+        self.setTextCursor(cursor)
+
+        return bubble_id
+
+    def clear(self):
+        """
+        Clears all bubbles from the editor.
+        """
+        for format in self._formats.values():
+            del format
+        self._bubble_text_object.clear()
+        super(BubbleEditWidget, self).clear()
+
+    def clear_typed_text(self):
+        """
+        Clears only typed text (not bubbles) from the editor.
+        :return:
+        """
+
+        text = self.toPlainText()
+        cursor = self.textCursor()
+
+        cursor.beginEditBlock()
+
+        # iterate over all characters in the editor. remove any that aren't the
+        # special replacement character. iterate backwards so that indices don't change
+        for i in reversed(range(0, len(text))):
+
+            if text[i] == self._OBJECT_REPLACEMENT_CHAR:
+                continue
+
+            cursor.setPosition(i, QtGui.QTextCursor.MoveAnchor)
+            cursor.movePosition(QtGui.QTextCursor.Right, QtGui.QTextCursor.KeepAnchor)
+            cursor.removeSelectedText()
+
+        cursor.endEditBlock()
+
+    def eventFilter(self, object, event):
+        """
+        Attempts to identify clicks on one of the editor's bubble widget's
+        remove button.
+
+        :param object: The observed object.
+        :type object: :class:`~PySide.QtCore.QObject`
+        :param event: The event to filter.
+        :type object: :class:`~PySide.QtCore.Qt.QEvent`
+        :return: True'' if the event was filtered, ''False'' otherwise.
+        """
+        if not isinstance(event, QtGui.QMouseEvent):
+            # only pass on mouse events
+            return False
+
+        # @todo:
+        # can't seem to figure out how to map the viewport position to the
+        # scroll area position. tried all combinations of mapto/from...
+        # resulting to just adding the scroll values. please fix if you can
+        edit_pos = QtCore.QPoint(
+            event.pos().x() + self.horizontalScrollBar().value(),
+            event.pos().y() + self.verticalScrollBar().value()
+        )
+
+        # for mouse events find the actual widget at the position
+        doc = self.document()
+        cursor_pos = doc.documentLayout().hitTest(edit_pos, QtCore.Qt.ExactHit)
+        format = doc.documentLayout().format(cursor_pos)
```

Comment:
Same here as before. I'd try to stay away from variable names that match builtins.